### PR TITLE
IB: TRT is not updated on ADD SYSTEM VERSIONING [fixes #413]

### LINKED ISSUE #413: TRT is not updated on ADD SYSTEM VERSIONING
### Reproduce
```sql
set @@system_versioning_alter_history= keep;
create or replace table t (x int) engine innodb;
insert into t values (1);
alter table t add system versioning;
select sys_trx_start from t into @trx_start;
select * from mysql.transaction_registry where transaction_id = @trx_start;
```
### Result
No records.
### Expected
TR record is found.

### PR DESCRIPTION
**Pull request by kevgs**

### CODE REVIEW

**Review by midenok — CHANGES_REQUESTED** [review]
TRT should not be updated when there are no records.

**Comment by kevgs** [comment]
Updated.

**Comment by midenok** [comment]
Please, rebase on latest trunk.

**Comment by kevgs** [comment]
Done.

File: storage/innobase/row/row0merge.cc [hunk 1]
```diff
@@ -1745,6 +1745,7 @@ row_merge_read_clustered_index(
 	char			new_sys_trx_start[8];
 	char			new_sys_trx_end[8];
 	byte			any_autoinc_data[8] = {0};
+	bool			insert_to_transaction_registry = false;
```
Comment (midenok) [inline storage/innobase/row/row0merge.cc]: bool vers_update_trt;

File: storage/innobase/handler/ha_innodb.cc [hunk 2]
```diff
@@ -3633,6 +3644,11 @@ static ulonglong innodb_prepare_commit_versioned(THD* thd, ulonglong *trx_id)
 	if (const trx_t* trx = thd_to_trx(thd)) {
 		*trx_id = trx->id;
 
+		if (trx->id && thd_alter_add_system_versioning(thd) && trx->mod_tables.size() == 1
+		    && trx->mod_tables.begin()->second.is_trx_versioned()) {
+			return get_new_trx_id_locked();
+		}
```
Comment (midenok) [inline storage/innobase/handler/ha_innodb.cc]: This check is worse than just checking bool flag.

Comment (kevgs) [reply storage/innobase/handler/ha_innodb.cc]: Well, probably. Is it just an opinion or I should fix it somehow?

**Comment by kevgs** [comment]
Updated.

File: storage/innobase/handler/ha_innodb.cc (hunk 2 above)
Comment (midenok) [reply storage/innobase/handler/ha_innodb.cc]: Update `mod_tables` such way that there would be no need to add this check.

**Comment by kevgs** [comment]
Updated.

File: storage/innobase/handler/ha_innodb.cc [hunk 3]
```diff
@@ -3638,7 +3638,9 @@ static ulonglong innodb_prepare_commit_versioned(THD* thd, ulonglong *trx_id)
 		     t != trx->mod_tables.end(); t++) {
 			if (t->second.is_trx_versioned()) {
 				DBUG_ASSERT(t->first->versioned());
-				DBUG_ASSERT(trx->undo_no);
+				if (!thd_alter_add_system_versioning(thd)) {
+					DBUG_ASSERT(trx->undo_no);
+				}
```
Comment (midenok) [inline storage/innobase/handler/ha_innodb.cc]: Remove `DBUG_ASSERT(trx->undo_no)`.

File: sql/sql_class.cc [hunk 4]
```diff
@@ -1248,6 +1248,11 @@ extern "C" my_thread_id next_thread_id_noinline()
 }
 #endif
 
+extern "C" bool thd_alter_add_system_versioning(MYSQL_THD thd)
+{
+  return thd->lex->sql_command == SQLCOM_ALTER_TABLE &&
+         thd->lex->vers_get_info().with_system_versioning;
+}
```
Comment (midenok) [inline sql/sql_class.cc]: Remove `thd_alter_add_system_versioning`.

**Comment by kevgs** [comment]
Updated.
`handler0alter.cc` diff fixes a bug introduced in https://github.com/tempesta-tech/mariadb/commit/4bc268d40686fb94bb53c5bb0b85f4ea97d70fa6 There was no corresponding test case before this patch.

**Review by midenok — APPROVED** [review]

**Pull request merged**